From 1823332e71fb1e9ac597af40c39ce9132a511413 Mon Sep 17 00:00:00 2001
From: Jingyi Wang <wangjingyi11@huawei.com>
Date: Wed, 7 Feb 2024 16:22:42 +0800
Subject: [PATCH 548/960] downstream: KVM: arm64: Make use of TWED feature

For HCR_EL2, TWEDEn(bit[59]) decides whether TWED is enabled, and
when the configurable delay is enabled, TWEDEL (bits[63:60]) encodes
the minimum delay in taking a trap of WFE caused by the TWE bit in
this register as 2^(TWEDEL + 8) cycles.

We use two kernel parameters "twed_enable" and "twedel" to configure
the register.

Signed-off-by: Zengruan Ye <yezengruan@huawei.com>
Signed-off-by: Jingyi Wang <wangjingyi11@huawei.com>
Signed-off-by: lishusen <lishusen2@huawei.com>
Signed-off-by: Slim6882 <15605196882@163.com>
Signed-off-by: chenyi <chenyi211@h-partners.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/include/asm/kvm_arm.h     |  7 +++++++
 arch/arm64/include/asm/kvm_emulate.h | 21 +++++++++++++++++++++
 arch/arm64/include/asm/kvm_host.h    |  5 +++++
 arch/arm64/kvm/arm.c                 |  9 +++++++++
 4 files changed, 42 insertions(+)

diff --git a/arch/arm64/include/asm/kvm_arm.h b/arch/arm64/include/asm/kvm_arm.h
index 5e8a0835470d..04ff06978d00 100644
--- a/arch/arm64/include/asm/kvm_arm.h
+++ b/arch/arm64/include/asm/kvm_arm.h
@@ -14,6 +14,7 @@
 
 /* Hyp Configuration Register (HCR) bits */
 
+#define HCR_TWEDEN	(UL(1) << 59)
 #define HCR_TID5	(UL(1) << 58)
 #define HCR_DCT		(UL(1) << 57)
 #define HCR_ATA_SHIFT	56
@@ -74,6 +75,12 @@
 #define HCR_VM		(UL(1) << 0)
 #define HCR_RES0	((UL(1) << 48) | (UL(1) << 39))
 
+#ifdef CONFIG_ARM64_TWED
+#define HCR_TWEDEL_SHIFT	60
+#define HCR_TWEDEL_MAX		(UL(0xf))
+#define HCR_TWEDEL_MASK		(HCR_TWEDEL_MAX << HCR_TWEDEL_SHIFT)
+#endif
+
 /*
  * The bits we set in HCR:
  * TLOR:	Trap LORegion register accesses
diff --git a/arch/arm64/include/asm/kvm_emulate.h b/arch/arm64/include/asm/kvm_emulate.h
index 382adf6befb6..c277cf7855e2 100644
--- a/arch/arm64/include/asm/kvm_emulate.h
+++ b/arch/arm64/include/asm/kvm_emulate.h
@@ -130,6 +130,27 @@ static inline void vcpu_set_wfx_traps(struct kvm_vcpu *vcpu)
 	vcpu->arch.hcr_el2 |= HCR_TWI;
 }
 
+#ifdef CONFIG_ARM64_TWED
+static inline void vcpu_set_twed(struct kvm_vcpu *vcpu)
+{
+	if (!cpus_have_final_cap(ARM64_HAS_TWED))
+		return;
+
+	if (twed_enable) {
+		u64 delay = (u64)twedel;
+
+		delay = (delay > HCR_TWEDEL_MAX) ? HCR_TWEDEL_MAX : delay;
+		vcpu->arch.hcr_el2 |= HCR_TWEDEN;
+		vcpu->arch.hcr_el2 &= ~HCR_TWEDEL_MASK;
+		vcpu->arch.hcr_el2 |= (delay << HCR_TWEDEL_SHIFT);
+	} else {
+		vcpu->arch.hcr_el2 &= ~HCR_TWEDEN;
+	}
+}
+#else
+static inline void vcpu_set_twed(struct kvm_vcpu *vcpu) {}
+#endif
+
 static inline void vcpu_ptrauth_enable(struct kvm_vcpu *vcpu)
 {
 	vcpu->arch.hcr_el2 |= (HCR_API | HCR_APK);
diff --git a/arch/arm64/include/asm/kvm_host.h b/arch/arm64/include/asm/kvm_host.h
index a34f97bdfc34..4717c0ca80b7 100644
--- a/arch/arm64/include/asm/kvm_host.h
+++ b/arch/arm64/include/asm/kvm_host.h
@@ -1204,6 +1204,11 @@ void __init kvm_hyp_reserve(void);
 static inline void kvm_hyp_reserve(void) { }
 #endif
 
+#ifdef CONFIG_ARM64_TWED
+extern bool twed_enable;
+extern unsigned int twedel;
+#endif
+
 void kvm_arm_vcpu_power_off(struct kvm_vcpu *vcpu);
 bool kvm_arm_vcpu_stopped(struct kvm_vcpu *vcpu);
 
diff --git a/arch/arm64/kvm/arm.c b/arch/arm64/kvm/arm.c
index 519710106d50..5885ea605cd3 100644
--- a/arch/arm64/kvm/arm.c
+++ b/arch/arm64/kvm/arm.c
@@ -113,6 +113,13 @@ static int vcpu_req_reload_wfi_traps(const char *val, const struct kernel_param
 	mutex_unlock(&kvm_lock);
 	return err;
 }
+#ifdef CONFIG_ARM64_TWED
+bool twed_enable;
+module_param(twed_enable, bool, 0644);
+
+unsigned int twedel;
+module_param(twedel, uint, 0644);
+#endif
 
 int kvm_arch_vcpu_should_kick(struct kvm_vcpu *vcpu)
 {
@@ -1060,6 +1067,8 @@ int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu)
 		kvm_arm_setup_debug(vcpu);
 		kvm_arch_vcpu_ctxflush_fp(vcpu);
 
+		vcpu_set_twed(vcpu);
+
 		/**************************************************************
 		 * Enter the guest
 		 */
-- 
2.43.0

